compiler/rustc_hir_typeck/src/method/probe.rs

```diff
@@ -950,51 +953,50 @@ impl<'a, 'tcx> ProbeContext<'a, 'tcx> {
-                        import_ids: import_ids.clone(),
-                        kind: TraitCandidate(new_trait_ref),
-                    },
-                    false,
-                );
-            });
-        } else {
-            debug_assert!(self.tcx.is_trait(trait_def_id));
-            if self.tcx.trait_is_auto(trait_def_id) {
-                return;
+        debug_assert!(self.tcx.is_trait(trait_def_id));
+        if self.tcx.trait_is_auto(trait_def_id) {
+            return;
+        }
+        for item in self.impl_or_trait_item(trait_def_id) {
+            // Check whether `trait_def_id` defines a method with suitable name.
+            if !self.has_applicable_self(&item) {
+                debug!("method has inapplicable self");
+                self.record_static_candidate(CandidateSource::Trait(trait_def_id));
+                continue;
             }
-            for item in self.impl_or_trait_item(trait_def_id) {
-                // Check whether `trait_def_id` defines a method with suitable name.
-                if !self.has_applicable_self(&item) {
-                    debug!("method has inapplicable self");
-                    self.record_static_candidate(CandidateSource::Trait(trait_def_id));
-                    continue;
-                }
 
-                let (xform_self_ty, xform_ret_ty) =
-                    self.xform_self_ty(&item, trait_ref.self_ty(), trait_substs);
-                self.push_candidate(
-                    Candidate {
-                        xform_self_ty,
-                        xform_ret_ty,
-                        item,
-                        import_ids: import_ids.clone(),
-                        kind: TraitCandidate(trait_ref),
-                    },
-                    false,
-                );
+            let (xform_self_ty, xform_ret_ty) =
+                self.xform_self_ty(&item, trait_ref.self_ty(), trait_substs);
+            self.push_candidate(
+                Candidate {
+                    xform_self_ty,
+                    xform_ret_ty,
+                    item,
+                    import_ids: import_ids.clone(),
+                    kind: TraitCandidate(trait_ref),
+                },
+                false,
+            );
+        }
+    }
+
+    fn assemble_extension_candidates_for_trait_or_alias(
+        &mut self,
+        import_ids: &SmallVec<[LocalDefId; 1]>,
+        trait_def_id: DefId,
+    ) {
+        debug!("assemble_extension_candidates_for_trait_or_alias(trait_def_id={:?})", trait_def_id);
+        if self.tcx.is_trait_alias(trait_def_id) {
+            for (predicate, _) in self.tcx.super_predicates_of(trait_def_id).predicates {
+                if let Some(poly_trait_predicate) = predicate.to_opt_poly_trait_pred() {
+                    self.assemble_extension_candidates_for_trait_or_alias(
+                        import_ids,
+                        poly_trait_predicate.def_id(),
+                    );
+                }
```

Comment:
I think this is the only change you need. Can you just replace the old call to `self.elaborate_bounds` with this, and not split up `assemble_extension_candidates_for_trait_or_alias`?

Reply:
Additionally, `super_predicates_of` has a quirk where, for trait aliases, it's defined to be *all* where clauses on the trait. You need to check that the self type of the predicate is `Self`, i.e. [`tcx.types.self_param`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.CommonTypes.html#structfield.self_param).

Reply:
But what if a trait alias has another trait alias amongst its bounds?  Don't we want to recursively assemble candidates from all explicit trait alias bounds?

Reply:
Perhaps I should add a test for that.

Reply:
Yeah, that looks perfect!